### PR TITLE
W6 l1, added loader spinner, category page, improved main page

### PR DESCRIPTION
**Pull request by sardorboboshov**

### CODE REVIEW

**Comment by vokanred** [comment]
What is this?

**Comment by sardorboboshov** [comment]
> What is this?

pr for 1st homework of previous week.

**Comment by vokanred** [comment]
Do you need the previous PR https://github.com/sardorboboshov/budgetify-w1/pull/7?
Or I should review only this?

**Comment by sardorboboshov** [comment]
> Do you need the previous PR #7? Or I should review only this?

you can only review this too: it includes the 2nd home task of week5 and 1st home task of 6th week.

**Comment by sardorboboshov** [comment]
> > Do you need the previous PR #7? Or I should review only this?
> 
> you can only review this too: it includes the 2nd home task of week5 and 1st home task of 6th week(previous week)

**Comment by vokanred** [comment]
Pls merge the previous PR and update this branch. Then I'll review

**Comment by vokanred** [comment]
You should merge master branch to this branch

**Comment by sardorboboshov** [comment]
> You should merge master branch to this branch

I kinda didnt understand that, should I open a new PR comparing that branch with master branch?

**Comment by vokanred** [comment]
Switch locally to w6_l1. Then `git pull origin master`. Then resolve conflicts. And then push changes `git push origin w6_l1`

**Comment by sardorboboshov** [comment]
> Switch locally to w6_l1. Then `git pull origin master`. Then resolve conflicts. And then push changes `git push origin w6_l1`.
> 
> Now w6_l1 doesn't have actual code from master.

Thanks for the guidance. I did exactly as you said, if something is wrong, please let me know

**Comment by sardorboboshov** [comment]
> Thanks for the guidance. I did exactly as you said, if something is wrong, please let me know

Thanks for the guidance. I did exactly as you said, if something is wrong, please let me know

File: client/src/app/main/accounts/accounts.component.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  styleUrls: ['./accounts.component.scss']
+})
+export class AccountsComponent implements OnInit {
+  @Output() accountIdx = new EventEmitter<number>();
+  data!: IUser;
+  accounts: IAccount[] = [];
+  selectedAccountIdx: number = 0;
+  constructor(private mainService: MainService) {}
+  ngOnInit() {
+    this.accountIdx.emit(this.selectedAccountIdx);
```
Comment (vokanred) [inline client/src/app/main/accounts/accounts.component.ts:17]: It's strange to emit in ngOnInit(). Why do you need this?

If you really need it just decorate  selectedAccountIdx: number = 0; as @Input and pass initial value from parent.
But I think it's not necessary.

File: client/src/app/main/transactions/transactions.component.html [hunk 2]
```diff
@@ -0,0 +1,68 @@
+<section class="wrapper">
```
Comment (vokanred) [inline client/src/app/main/transactions/transactions.component.html:1]: I think this component should be split into smaller components

File: client/src/app/main/transactions/transactions.component.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    if (
+      changes['selectedAccountIdx'].currentValue !==
+      changes['selectedAccountIdx'].previousValue
+    ) {
+      this.getTransActionsData();
+    }
+  }
+  getTransActionsData() {
+    const user_id = localStorage.getItem('userId');
+    this.mainService
```
Comment (vokanred) [inline client/src/app/main/transactions/transactions.component.ts:36]: async pipe or unsubscribe

File: client/src/app/main/transactions/transactions.component.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+      .getAllTransactionsData(Number(user_id), this.selectedAccountIdx)
+      .subscribe((res: any) => {
+        this.transActionsData = res.transactions;
+      });
+    this.mainService
```
Comment (vokanred) [inline client/src/app/main/transactions/transactions.component.ts:41]: async pipe or unsubscribe

File: client/src/app/main/transactions/transactions.component.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+      .getAccountData(Number(user_id), this.selectedAccountIdx)
+      .subscribe((res: any) => {
+        this.currency = res.currency;
+      });
+  }
+  setTransActionsType(type: string) {
+    const user_id = localStorage.getItem('userId');
+    this.mainService
```
Comment (vokanred) [inline client/src/app/main/transactions/transactions.component.ts:49]: async pipe or unsubscribe

File: client/src/app/shared/services/spinner.service.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+})
+export class SpinnerService {
+  private isSpinnerVisible$: Subject<boolean> = new BehaviorSubject<boolean>(
+    false
+  );
+  constructor() {}
+  getIsSpinnerVisible$() {
+    return this.isSpinnerVisible$
+      .asObservable()
+      .pipe(observeOn(asyncScheduler));
```
Comment (vokanred) [inline client/src/app/shared/services/spinner.service.ts:15]: Did lector show you asyncScheduler?

File: controllers/categoryController.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+exports.createCategory = async (req, res) => {
+  try {
+    const { title, type } = req.body;
+    if (!title || !type) {
+      return res.json({ message: 'Title and type are required' });
+    }
+    const categories = await Category.find({});
+    categories.forEach((category_) => {
+      if (category_.title.toLowerCase() === title) {
```
Comment (vokanred) [inline controllers/categoryController.js:11]: You shouldn't check this manually in code.

Just add unique: true to your model and mongoose will throw error
Screen Shot 2022-04-14 at 10 06 11 PM
.

Comment (vokanred) [reply controllers/categoryController.js:11]: I see that you've already added it. So you can remove this check

**Review by vokanred — CHANGES_REQUESTED** [review]
Nice.
Pls review comments and make changes

File: client/src/app/layout/header/header.component.ts [hunk 8]
```diff
@@ -1,15 +1,21 @@
...
-  constructor() { }
+  user_name: string = 'username';
+  constructor(private mainService: MainService) {}
 
   ngOnInit(): void {
+    this.setUser();
+  }
+  setUser() {
+    const user_id = localStorage.getItem('userId');
+    this.mainService.getData(Number(user_id)).subscribe((data: any) => {
```
[the remark this answers is not in the document]
Comment (sardorboboshov) [reply client/src/app/layout/header/header.component.ts:17]: image
image

username is not being displayed, also there is no "it is running" showing on the console.
can you give some advice how can I use async pipe here

File: client/src/app/shared/services/spinner.service.ts (hunk 6 above)
Comment (sardorboboshov) [reply client/src/app/shared/services/spinner.service.ts:15]: > Did lector show you asyncScheduler?

yes

File: client/src/app/main/transactions/transactions.component.ts (hunk 4 above)
Comment (sardorboboshov) [reply client/src/app/main/transactions/transactions.component.ts:41]: > async pipe or unsubscribe

if i use async pipe for currency and use that in template it is giving me that error:
image
image

File: client/src/app/main/transactions/transactions.component.ts (hunk 5 above)
Comment (sardorboboshov) [reply client/src/app/main/transactions/transactions.component.ts:49]: > async pipe or unsubscribe

if i use async pipe I can only display transactions on page and I wont be able to access that on .ts file therefore I cannt filter transactions according to types of transactions. What do u suggest me in that case?

**Comment by sardorboboshov** [comment]
> Nice. Pls review comments and make changes

I commited changes according to your feedbacks. 
1)Used pipe - async apart from transaction components where I needed to access and filter data
2)Unsubscribed on transaction component
3)Splitted transaction component into two component
4)Resolved .gitignore file, removed checking on creating category.
If you not seeing them or in the case of any error, please let me know.

File: client/src/app/main/accounts/accounts.component.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    this.setAccount();
+  }
+  setAccount() {
+    const user_id = localStorage.getItem('userId');
+    this.mainService
```
[the remark this answers is not in the document]
Comment (vokanred) [reply client/src/app/main/accounts/accounts.component.ts:22]: this.accounts$ = this.mainService.getAllAccountsData(Number(user_id))

File: client/src/app/main/category/category.component.ts [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+  selector: 'app-category',
+  templateUrl: './category.component.html',
+  styleUrls: ['./category.component.scss']
+})
+export class CategoryComponent implements OnInit {
+  categories$: Observable<ICategory[]> = new Observable<[]>();
+  constructor(private mainService: MainService) {}
+
+  ngOnInit(): void {
+    this.mainService.getAllCategories().pipe((res) => (this.categories$ = res));
```
Comment (vokanred) [inline client/src/app/main/category/category.component.ts:15]: this.categories$ = this.mainService.getAllCategories()

**Review by vokanred — CHANGES_REQUESTED** [review]
Almost done. Just fix all places where you've added async pipe.
I've written couple example how it should look

**Comment by sardorboboshov** [comment]
> Almost done. Just fix all places where you've added async pipe. I've written couple example how it should look

fixed all places where async pipe is used. if you not seeing them, please let me know.

**Review by vokanred — APPROVED** [review]
Approved

**Pull request merged**